tests/lib/onoff/src/main.c

```diff
@@ -0,0 +1,1164 @@
+	rc = onoff_service_init(&srv, start, stop, reset,
+				ONOFF_SERVICE_START_SLEEPS);
+	zassert_equal(rc, 0,
+		      "service init");
+	start_state.async = true;
+	start_state.retval = 12;
+
+	struct isr_call_state isr_state = {
+		.srv = &srv,
+		.cli = &spinwait_cli,
+	};
+	struct k_timer timer;
+
+	init_spinwait(&spinwait_cli);
+	k_timer_init(&timer, isr_request, NULL);
+	k_timer_user_data_set(&timer, &isr_state);
+
+	k_timer_start(&timer, K_MSEC(1), K_NO_WAIT);
+	rc = k_sem_take(&isr_sync, K_MSEC(10));
+	zassert_equal(rc, 0,
+		      "isr sync");
+
+	zassert_equal(isr_state.result, -EWOULDBLOCK,
+		      "isr request");
+	zassert_equal(cli_result(&spinwait_cli), -EAGAIN,
+		      "isr request result");
+}
+
+static void test_sync(void)
+{
+	int rc;
+	struct onoff_service srv;
+
+	clear_transit();
+
+	rc = onoff_service_init(&srv, start, stop, reset, 0);
+	zassert_equal(rc, 0,
+		      "service init");
+
+	/* WHITEBOX: request that triggers on returns positive */
+	init_spinwait(&spinwait_cli);
+	rc = onoff_request(&srv, &spinwait_cli);
+	zassert_equal(rc, 2,    /* WHITEBOX starting request */
+		      "req ok");
+	zassert_equal(srv.refs, 1U,
+		      "reset req refs: %u", srv.refs);
+
+	init_spinwait(&spinwait_cli);
+	rc = onoff_request(&srv, &spinwait_cli);
+	zassert_equal(rc, 0,    /* WHITEBOX on request */
+		      "req ok");
+	zassert_equal(srv.refs, 2U,
+		      "reset req refs: %u", srv.refs);
+
+	init_spinwait(&spinwait_cli);
+	rc = onoff_release(&srv, &spinwait_cli);
+	zassert_equal(rc, 1,    /* WHITEBOX non-stopping release */
+		      "rel ok");
+	zassert_equal(srv.refs, 1U,
+		      "reset rel refs: %u", srv.refs);
+
+	init_spinwait(&spinwait_cli);
+	rc = onoff_release(&srv, &spinwait_cli);
+	zassert_equal(rc, 2,    /* WHITEBOX stopping release*/
+		      "rel ok: %d", rc);
+	zassert_equal(srv.refs, 0U,
+		      "reset rel refs: %u", srv.refs);
+
+	init_spinwait(&spinwait_cli);
+	rc = onoff_release(&srv, &spinwait_cli);
+	zassert_equal(rc, -EALREADY,
+		      "rel noent");
+}
+
+static void test_async(void)
+{
+	int rc;
+	struct onoff_service srv;
+	struct k_poll_signal sig[2];
+	struct onoff_client cli[2];
+	unsigned int signalled = 0;
+	int result = 0;
+
+	clear_transit();
+	start_state.async = true;
+	start_state.retval = 23;
+	stop_state.async = true;
+	stop_state.retval = 17;
+
+	rc = onoff_service_init(&srv, start, stop, reset,
+				ONOFF_SERVICE_START_SLEEPS
+				| ONOFF_SERVICE_STOP_SLEEPS);
+	zassert_equal(rc, 0,
+		      "service init");
+
+	/* WHITEBOX: request that triggers on returns positive */
+	init_notify_sig(&cli[0], &sig[0]);
+	rc = onoff_request(&srv, &cli[0]);
+	zassert_equal(rc, 2,    /* WHITEBOX starting request */
+		      "req ok");
+	k_poll_signal_check(&sig[0], &signalled, &result);
+	zassert_equal((bool)signalled, false,
+		      "cli signalled");
+	zassert_equal(srv.refs, 0U,
+		      "reset req refs: %u", srv.refs);
+
+
+	/* Non-initial request from ISR is OK */
+	struct onoff_client isrcli;
```

Comment:
I am surprised checkpatch did not raised a warning/error here. Never declare a variable in a middle of a code block

Reply:
There's lots of precedent for it in Zephyr, and the feature was added twenty years ago in C99 which Zephyr implicitly requires.  Unless there's a documented coding standard that disallows this I'm going to keep it in place.

Reply:
ok, after all even basic checkpatch warnings did not stop other PRs to get merged.